packages/shared/lib/core/wallet/classes/activity.class.ts

```diff
@@ -96,153 +43,60 @@ export class Activity implements IActivity {
 
-            this.isAsync = isOutputAsync(output)
-            this.isClaimed = false
-            this.asyncStatus = this.isAsync ? this.getAsyncStatus(new Date()) : null
-
-            const nativeToken = getNativeTokenFromOutput(output)
-
-            this.assetId = nativeToken?.id ?? String(COIN_TYPE[get(activeProfile).networkProtocol])
-            const asset = getPersistedAsset(this.assetId)
-            this.isAssetHidden = !asset || asset.hidden
-            this.isRejected = isActivityHiddenForAccountId(account.id, transactionId)
-
-            const { storageDeposit, giftedStorageDeposit } = getStorageDepositFromOutput(output)
-            this.storageDeposit = storageDeposit
-            this.giftedStorageDeposit = giftedStorageDeposit
-            this.rawAmount = nativeToken
-                ? Number(nativeToken?.amount)
-                : getAmountFromOutput(output) - this.storageDeposit
-
-            this.metadata = getMetadataFromOutput(output)
-            this.tag = getTagFromOutput(output)
-            this.expirationDate = getExpirationDateFromOutput(output)
+        if (type === ActivityType.Transaction) {
+            this.data = getTransactionActivityData(processedTransaction, account)
         } else if (type === ActivityType.Foundry) {
-            const { output, outputIndex } = getFoundryOutputFromTransaction(outputs)
-            this.id = outputIdFromTransactionData(transactionId, outputIndex)
-            this.outputId = outputIdFromTransactionData(transactionId, outputIndex) // not required for foundry activities
-
-            const nativeToken = getNativeTokenFromOutput(output)
-
-            this.assetId = nativeToken?.id ?? String(COIN_TYPE[get(activeProfile).networkProtocol])
-            const asset = getPersistedAsset(this.assetId)
-            this.isAssetHidden = !asset || asset.hidden
-
-            const { storageDeposit, giftedStorageDeposit } = getStorageDepositFromOutput(output) // probably we need to sum up all storage deposits
-            this.storageDeposit = storageDeposit
-            this.giftedStorageDeposit = giftedStorageDeposit
-            this.rawAmount = nativeToken
-                ? Number(nativeToken?.amount)
-                : getAmountFromOutput(output) - this.storageDeposit
+            this.data = getFoundryActivityData(processedTransaction)
         }
     }
 
     updateFromPartialActivity(partialActivity: Partial<IActivity>): void {
         Object.assign(this, partialActivity)
     }
 
+    updateDataFromPartialActivity(
+        partialData: IPartialTransactionActivityDataWithType | IPartialFoundryActivityDataWithType
+    ): void {
+        if (partialData.type === ActivityType.Transaction && this.data.type === ActivityType.Transaction) {
+            this.data = { ...this.data, ...partialData }
+        } else if (partialData.type === ActivityType.Foundry && this.data.type === ActivityType.Foundry) {
+            this.data = { ...this.data, ...partialData }
+        }
```

Comment:
I'm not sure whether the conditionals are necessary here. Both branches do the same thing, right?

Reply:
initially i had the check:
`if (partialData.type === this.data.type`  to assure that we don't set foundry data on a transaction or vice versa. But then typescript is angry because we have `data` and `partialData` which both can be `TypeA | TypeB`. With this implementation we narrow down both of them and assure typescript, that `TypeA` is only assigned to `TypeA`